### ISSUE TITLE
Generate MIgration.cs and apply in runtime ? Enterpise scenario!

### ISSUE THREAD

**Issue by Diaskhan** [dotnet/efcore#11819]
Hello friends I got very deep question about migrations and roslyn.

I have this code !

My scenario, that i want to generate class model for example DefinitionProcess
Then by this code is creating migration CS file.
The question i's Could I migrate database from object in memory or via generated code by scaffolder.ScaffoldMigration("MyMigration","MyApp.Data");

The case is that functional aming to Enterpise systems !
Or maybe Ef core provide change scheme on the fly ?
Like create table, alter column, without using dotnet update-database ?

```

//Program.cs 
using Microsoft.EntityFrameworkCore.Infrastructure;
using Microsoft.EntityFrameworkCore.Internal;
using Microsoft.EntityFrameworkCore.Migrations;
using Microsoft.EntityFrameworkCore.Storage;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.EntityFrameworkCore.Migrations.Design;
using System;
using System.IO;
using Microsoft.EntityFrameworkCore.Design.Internal;
using Microsoft.EntityFrameworkCore.Design;
using Models;
using Microsoft.EntityFrameworkCore.Migrations.Internal;

namespace db_migrator1
{
    class Program
    {
        static void Main(string[] args)
        {

            using (var db = new MyDBcontext())
            {
                var reporter = new OperationReporter(
                    new OperationReportHandler(
                        m => Console.WriteLine("  error: " + m),
                        m => Console.WriteLine("   warn: " + m),
                        m => Console.WriteLine("   info: " + m),
                        m => Console.WriteLine("verbose: " + m)));

                var designTimeServices = new ServiceCollection()
                    .AddSingleton(db.GetService<IHistoryRepository>())
                    .AddSingleton(db.GetService<IMigrationsIdGenerator>())
                    .AddSingleton(db.GetService<IMigrationsModelDiffer>())
                    .AddSingleton(db.GetService<IMigrationsAssembly>())
                    .AddSingleton(db.Model)
                    .AddSingleton(db.GetService<ICurrentDbContext>())
                    .AddSingleton(db.GetService<IDatabaseProvider>())
                    .AddSingleton<MigrationsCodeGeneratorDependencies>()
                    .AddSingleton<ICSharpHelper, CSharpHelper>()
                    .AddSingleton<CSharpMigrationOperationGeneratorDependencies>()
                    .AddSingleton<ICSharpMigrationOperationGenerator, CSharpMigrationOperationGenerator>()
                    .AddSingleton<CSharpSnapshotGeneratorDependencies>()
                    .AddSingleton<ICSharpSnapshotGenerator, CSharpSnapshotGenerator>()
                    .AddSingleton<CSharpMigrationsGeneratorDependencies>()
                    .AddSingleton<IMigrationsCodeGenerator, CSharpMigrationsGenerator>()
                    .AddSingleton<IOperationReporter>(reporter)
                    .AddSingleton<MigrationsScaffolderDependencies>()
                    .AddSingleton<MigrationsScaffolder>()
                    .AddSingleton<ISnapshotModelProcessor, SnapshotModelProcessor>()
                    .BuildServiceProvider();

                var scaffolder = designTimeServices.GetRequiredService<MigrationsScaffolder>();

                var migration = scaffolder.ScaffoldMigration(
                    "MyMigration",
                    "MyApp.Data");

                File.WriteAllText(
                    migration.MigrationId + migration.FileExtension,
                    migration.MigrationCode);
                File.WriteAllText(
                    migration.MigrationId + ".Designer" + migration.FileExtension,
                    migration.MetadataCode);
                File.WriteAllText(migration.SnapshotName + migration.FileExtension,
                    migration.SnapshotCode);
            }


        }
    }
}

//DefinitionProcess.cs 
using System;
using System.ComponentModel.DataAnnotations.Schema;

namespace Models
{
    public class DefinitionProcess
    {
        public long Id { get; set; }
        public string Name { get; set; } //Название процесса
        public int Ver { get; set; } //версия
        public DateTime DateDeployed { get; set; } //дата заливки процесса

        //Добавим позже
        //--Author - User_id
        //image - blob


    }
}

//MyDBcontext.cs 
using Microsoft.EntityFrameworkCore;
using Models;


namespace Models
{
    public class MyDBcontext:DbContext
    {
        public MyDBcontext() { }
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            base.OnConfiguring(optionsBuilder);
            optionsBuilder.UseNpgsql("User Id=postgres;Server=localhost;port=5432;Database=Falcon111;Pooling=true");
        }
        public MyDBcontext(DbContextOptions<MyDBcontext> options) : base(options) { }
        public DbSet<DefinitionProcess> DefinitionProcess { get; set; }

    }
}



/////Generated CS file migration 
// <auto-generated />
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Infrastructure;
using Microsoft.EntityFrameworkCore.Metadata;
using Microsoft.EntityFrameworkCore.Migrations;
using Microsoft.EntityFrameworkCore.Storage;
using Microsoft.EntityFrameworkCore.Storage.Internal;
using Models;
using System;

namespace MyApp.Data.Migrations
{
    [DbContext(typeof(MyDBcontext))]
    [Migration("20180424095456_MyMigration")]
    partial class MyMigration
    {
        protected override void BuildTargetModel(ModelBuilder modelBuilder)
        {
#pragma warning disable 612, 618
            modelBuilder
                .HasAnnotation("Npgsql:ValueGenerationStrategy", NpgsqlValueGenerationStrategy.SerialColumn)
                .HasAnnotation("ProductVersion", "2.0.1-rtm-125");

            modelBuilder.Entity("Models.DefinitionProcess", b =>
                {
                    b.Property<long>("Id")
                        .ValueGeneratedOnAdd();

                    b.Property<DateTime>("DateDeployed");

                    b.Property<string>("Name");

                    b.Property<int>("Ver");

                    b.HasKey("Id");

                    b.ToTable("DefinitionProcess");
                });
#pragma warning restore 612, 618
        }
    }
}

```

**Comment by ajcvickers** [comment]
@Diaskhan Updating the database from application code is something we discourage. The main reason for this is that in general it is not safe to apply migrations to a running application without ensuring that all access to the database from all requests/threads/processes/servers/etc is blocked during the migration process.

That being said, if you are in full control of the environment such that you are sure it is safe, then it can be done, but even then I would discourage against trying to include automatic generation of the migrations in this process since you will not be able to review the scaffolded migration for correctness. You'll also have to handle the model snapshot in some way, which can be tricky, and will be even more tricky where multiple developers are working on the project.

So bottom line is that the best advice in this case is to not go down this path.
